### PR TITLE
Improve /errlog JSON escaping logic and add tests

### PR DESCRIPTION
**Pull request by nguerrera**
Warming up for pending update to latest SARIF draft by fixing #9769 first

1. Stop escaping forward slashes (optional per spec and users don't want it)
2. Add tests converted from DataContractSerializer tests that stress the code we borrowed from there for escaping.
3. Let the underlying TextWriter handle all newlines

@dotnet/roslyn-compiler @lgolding @srivatsn  @mavasani @michaelcfanning

### CODE REVIEW

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        // #5. 0xFFFE - 0xFFFF
+        [InlineData(@"\ufffe", (char)0xfffe)]
+        [InlineData(@"\uffff", (char)0xffff)]
+        public static void Character(string expected, char value)
+        {
+            WriteInMultiplePositionsAndCheck(expected, value.ToString());
+        }
+
+        [Theory]
+        [InlineData(@"\ud83d\udca9", "💩")]
```
Comment (lgolding) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs]: Srsly, that's the emoji you chose? ;-)

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        [InlineData("1", 1)]
+        [InlineData("42", 42)]
+        [InlineData("2147483647", int.MaxValue)]
+        public static void Integers(string expected, int value)
+        {
+            Assert.StrictEqual(expected, WriteToString(value));
+        }
+
+        
+        [Theory]
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs]: I have a theory that this is the most heavy use of theory in our code base. ;)

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (lgolding) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs]: Curious about this file. Why do we need to unit test `JsonWriter`?

Comment (lgolding) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs]: Oh, sorry, I forgot it was _our_ `JsonWriter` and not Json.NET's :)

**Comment by lgolding** [comment]
:+1:

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs (hunk 1 above)
Comment (nguerrera) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/JsonWriterTests.cs]: Changed it to thumbs up. 😉

**Comment by nguerrera** [comment]
@dotnet/roslyn-compiler: Is this good to go?

**Comment by nguerrera** [comment]
Rebased in to #11233

**Pull request closed without merging**